nodejs/awsx/apigateway/authorizer.ts

```diff
@@ -0,0 +1,355 @@
+ * @param args - configuration information for the token Lambda.
+ */
+export function getTokenLambdaAuthorizerDefinition(args: TokenAuthorizerArgs): LambdaAuthorizerDefinition {
+    return {
+        authorizerName: args.authorizerName,
+        parameterName: args.header || "Authorization",
+        parameterLocation: "header",
+        authType: "oauth2",
+        type: "token",
+        handler: args.handler,
+        identityValidationExpression: args.identityValidationExpression,
+        authorizerResultTtlInSeconds: args.authorizerResultTtlInSeconds,
+    };
+}
+
+/**
+ * The set of arguments for constructing a request LambdaAuthorizerDefinition resource.
+ */
+export interface RequestAuthorizerArgs {
+
+    /**
+     * The name for the Authorizer to be referenced as. This must be unique for each unique authorizer
+     * within the API. If no name if specified, a name will be generated for you.
+     */
+    authorizerName?: string;
+
+    /**
+     * queryParameters is an array of the expected query parameter keys used to authorize a request.
+     * While this argument is optional, at least one queryParameter or one header must be defined.
+     * */
+    queryParameters?: string[];
+
+    /**
+     * headers is an array of the expected header keys used to authorize a request.
+     *  While this argument is optional, at least one queryParameter or one header must be defined.
+     * */
+    headers?: string[];
+
+    /**
+     * The authorizerHandler specifies information about the authorizing Lambda. You can either set
+     * up the Lambda separately and just provide the required information or you can define the
+     * Lambda inline using a JavaScript function.
+     */
+    handler: LambdaAuthorizerInfo | aws.lambda.EventHandler<AuthorizerEvent, AuthorizerResponse>;
+
+    /**
+     * The number of seconds during which the resulting IAM policy is cached. Default is 300s. You
+     * can set this value to 0 to disable caching. Max value is 3600s. Note - if you are sharing an
+     * authorizer across more than one route you will want to disable the cache or else it will
+     * cause problems for you.
+     */
+    authorizerResultTtlInSeconds?: number;
+}
+
+/**
+ * getRequestLambdaAuthorizerDefinition is a helper function to generate a request
+ * LambdaAuthorizerDefinition.
+ * @param name - the name for the authorizer. This must be unique for each unique authorizer in the
+ * API.
+ * @param args - configuration information for the token Lambda.
+ */
+export function getRequestLambdaAuthorizerDefinition(args: RequestAuthorizerArgs): LambdaAuthorizerDefinition {
+    let parameterName: string;
+    let location: "header" | "query";
+
+    const numQueryParams = getLength(args.queryParameters);
+    const numHeaders = getLength(args.headers);
+
+    if (numQueryParams === 0 && numHeaders === 0) {
+        throw new Error("Must specify at least one parameter");
```

Comment:
error message is a bit vague.  it can be useful to be more specific.  like `"[args.queryParameters] and [args.headers] were both empty.  At least one query parameter or header must be specified"`.